contracts/CDPLeverage.sol

```diff
@@ -0,0 +1,456 @@
+            uint256 remainingDai;
+            uint256 excessWeth;
+
+            IWETH(weth).deposit.value(msg.value)();
+
+            uint256 mkrFee;
+            uint256 wethFee;
+            (mkrFee, wethFee) = wipeFee(sender.cdpID, remainingDebt);
+
+            // Pay for governance fee
+            require(wethFee <= msg.value, "Not enough ether provided for fees");
+            dex.buyAllAmount(gov, mkrFee, weth, wethFee);
+            excessWeth = sub(msg.value,wethFee);
+
+            // wipe last layer of cdp
+            remainingDebt = wipeDebt(sender.cdpID, daiAmount, remainingDebt);
+            releasedPeth = wdiv(wmul(daiAmount,makerLR),sender.purchPrice);
+            releaseWeth(sender.cdpID, releasedPeth);
+            remainingPeth -= releasedPeth;
+
+            // empty remaining debt or reach dust pan condition (CDP peth = 0.0051), whichever comes first
+            while (remainingDebt > 0) {
+                daiAmount = marketBuy(dai, weth, releasedPeth);
+
+                if (daiAmount > remainingDebt) {
+                    remainingDai = sub(daiAmount,remainingDebt);
+                    daiAmount = remainingDebt;
+                }
+
+                remainingDebt = wipeDebt(sender.cdpID, daiAmount, remainingDebt);
+                releasedPeth = wdiv(wmul(daiAmount,makerLR),sender.priceFloor);
+
+                // dust pan condition, where remainingPeth cannot be below 0.005 peth
+                if (sub(remainingPeth,releasedPeth) < 5000000000000000) {
+                    if (remainingDebt == 0) {
+                        releasedPeth = remainingPeth;
+                    } else {
+                        releasedPeth = sub(remainingPeth,5100000000000000);
+                        releaseWeth(sender.cdpID, releasedPeth);
+                        remainingPeth = 5100000000000000;
+                        break;
+                    }
+                }
+
+               releaseWeth(sender.cdpID, releasedPeth);
+               remainingPeth -= releasedPeth;
+            }
+
+            // release any final peth
+            uint256 finalPeth;
+            if (remainingDebt == 0 && remainingPeth != 0) {
+                finalPeth = tub.ink(sender.cdpID);
+                releaseWeth(sender.cdpID, finalPeth);
+            } else {
+                finalPeth = 0;
+            }
+
+            // convert left over dai to weth
+            wethAmount = marketBuy(weth, dai, remainingDai);
+
+            payout = add(add(add(releasedPeth,wethAmount),finalPeth),excessWeth);
+        }
+
+        // convert WETH to ETH
+        IWETH(weth).withdraw(payout);
+
+        // give cdp to owner
+        // if dustpan condition hit, cdp will have 0.0051 peth left
+        tub.give(sender.cdpID, msg.sender);
+
+        msg.sender.transfer(payout);
+
+       // delete the sender information from investorAddresses array
+        deleteEntity(msg.sender);
+
+        emit ClosePosition(msg.sender, currPriceEth);
```

Comment:
Could we log this position's principal and the eth value redeemed from liquidating as well?